### PR TITLE
feat(azure): support service principals and managed identities

### PR DESCRIPTION
**Pull request by sanderaernouts**
## Hey, I just made a Pull Request!

Azure DevOps recently added support for authenticating with a service principal or managed identity instead of a personal access token (see this [link](https://devblogs.microsoft.com/devops/introducing-service-principal-and-managed-identity-support-on-azure-devops/)). This PR adds support for service principals and managed identities to the azure (devops) integration. Note that service principal and managed identity authentication is only available on Azure AD backed Azure DevOps organizations.

This will also help with supporting multiple azure organizations (#10431) as long as they are backed by the same Azure AD. It seems that it is also possible to use [a managed identity from a different Azure AD tenant](https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/service-principal-managed-identity?view=azure-devops&preserve-view=true#q-can-i-add-a-managed-identity-from-a-different-tenant-to-my-organization) but that requires more code (I think) and supporting multiple Azure DevOps organizations is not the goal of this PR.

I have tested this with an Azure AD backed Azure DevOps org and it works as expected.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] ~~Screenshots attached (for UI changes)~~
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by sanderaernouts** [comment]
I still have to do some testing with an Azure AD backed Azure DevOps instance but the changes can be reviewed  already.

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for the contribution @sanderaernouts, this is a very nice addition.

I'm one of the authors of the Azure DevOps plugin and I often review related changes. I also tend to test things with out Azure DevOps Server instance to be sure it still works for on-prem users.

I left a few comments, mostly around the communication and documentation side of things, the code overall looks good to me but will let the maintainers dive into that more

File: .changeset/orange-taxis-exercise.md
Comment (sanderaernouts) [reply .changeset/orange-taxis-exercise.md:8]: Haven't worked with Azure DevOps Server for a while but this feature might not make sense there. If I remember correctly on-prem installations are normally backed by an on-prem AD where service accounts can be used.

Public preview means, to my best knowlegde, that the feature is available to any customer but it is without an SLA. I'll mention the feature is currently in public preview 👍

File: docs/integrations/azure/locations.md
Comment (sanderaernouts) [reply docs/integrations/azure/locations.md]: Done ✅

File: docs/integrations/azure/locations.md
Comment (sanderaernouts) [reply docs/integrations/azure/locations.md]: Done ✅

**Review by awanlin — APPROVED** [review]

File: packages/integration/api-report.md
Comment (awanlin) [inline packages/integration/api-report.md:762]: You'll need to fix this as this is what is causing the build to fail

**Comment by awanlin** [comment]
Hi @sanderaernouts, I approved from my perspective, the maintainer will also want to look this over. 

Also, just noticed the build is failing and from the looks of it you have a warning in the API Report, I left a comment there

File: packages/integration/api-report.md
Comment (sanderaernouts) [reply packages/integration/api-report.md:762]: I missed that, will have a look 👍

Comment (sanderaernouts) [reply packages/integration/api-report.md]: The warnings should be gone now :+1:

**Comment by sanderaernouts** [comment]
@benjdlambert I tested this implementation with our Azure DevOps instance and fixed a small issue. It now works as expected 🎉 . So from my side it is good to merge after reviews.

File: packages/integration/src/azure/config.ts
Comment (benjdlambert) [inline packages/integration/src/azure/config.ts:47]: Shouldn't this be something like this?

```suggestion
  credential?: AzureCredential;
```

Also I wonder if we should actually just change this to `authentication` and move `{ token?: string }` to be part of the union and deprecate the top level one? @backstage/maintainers?

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Thanks for this, it's looking good I'm just not sure about the naming of `credential` it feels a little out of place I think looking at the current setup and wonder if theres a better way to do this. Feels like this pattern could be useful for the other integrations too, want to get some more feedback from other maintainers here too.

File: packages/integration/src/azure/config.ts
Comment (sanderaernouts) [reply packages/integration/src/azure/config.ts]: You are right, I did this because I used the type predicates inside `isValidCredential(...)` so it turns out typescript was not able to narrow the type to `ClientSecret | ManagedIdentity` after this `if`. Fixed it 👍 

I thought about making `token` part of the union but I looked at the `github` integration and there it's `token` or `apps` and this felt sort of similar. It would make the code simpler because then the type narrowing can replace some of the checks I added now.

**Comment by sanderaernouts** [comment]
> Thanks for this, it's looking good I'm just not sure about the naming of `credential` it feels a little out of place I think looking at the current setup and wonder if theres a better way to do this. Feels like this pattern could be useful for the other integrations too, want to get some more feedback from other maintainers here.

I went with `identity` first but that did not feel right either. I went with `credential` in the end because it is the credential you use to authenticate with Azure DevOps and it aligns with the terminology in the Azure identity SDK. Though it might not allign with the terminology in backstage so I'm open to suggestions :). 

I think `authentication` could work but then that would have to be applied to other integrations as well I think?
```yaml
integrations:
  azure:
    host: dev.azure.com
    authentication:
      token: ...
  github:
    host: github.com
    authentication:
      token: ....
```

```yaml
integrations:
  azure:
    host: dev.azure.com
    authentication:
      clientId: ...
      clientSecret: ...
      tenantId: ...
  github:
    host: github.com
    authentication:
      apps: ....
```

**Comment by sanderaernouts** [comment]
@benjdlambert agree, just wrote down an example of what it would look like for other integrations like `github`. I'll hold off on changing it to `authentication` until we get some more feedback on this.

**Comment by benjdlambert** [comment]
@sanderaernouts we chatted amongst ourselves in @backstage/maintainers, and we think that it's fine just to keep this as `credential` for now. Thinking that we might get to a place where we need the `authentication` block but this one requirement doesn't really force us to have to do this with all the other ones just yet and rather than start a migration effort for it, it's probably best to keep it as is.

**Review by benjdlambert — APPROVED** [review]
And with that 👍

**Pull request merged**
